src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java

```diff
@@ -0,0 +1,198 @@
+import com.google.cloud.bigquery.JobStatistics;
+import com.google.cloud.bigquery.JobStatus;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableResult;
+import io.cdap.cdap.api.metrics.Metrics;
+import io.cdap.cdap.etl.api.StageMetrics;
+import io.cdap.cdap.etl.api.action.ActionContext;
+
+import io.cdap.cdap.etl.mock.validation.MockFailureCollector;
+import io.cdap.plugin.gcp.bigquery.exception.BigQueryJobExecutionException;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+public class BigQueryExecuteTest {
+  @Mock
+  BigQuery bigQuery;
+  @Mock
+  Job queryJob;
+  @Mock
+  JobStatus jobStatus;
+  @Mock
+  BigQueryError bigQueryError;
+  @Mock
+  TableResult queryResults;
+  @Mock
+  JobStatistics.QueryStatistics queryStatistics;
+  @Mock
+  ActionContext context;
+  @Mock
+  StageMetrics stageMetrics;
+  @Mock
+  Metrics metrics;
+  QueryJobConfiguration queryJobConfiguration;
+  BigQueryExecute.Config config;
+  JobInfo jobInfo;
+  JobId jobId;
+  BigQueryExecute bq;
+  MockFailureCollector failureCollector;
+
+  @Before
+  public void setUp() throws InterruptedException, NoSuchMethodException {
+    MockitoAnnotations.initMocks(this);
+    failureCollector = new MockFailureCollector();
+    queryJobConfiguration = QueryJobConfiguration.newBuilder("select * from test").build();
+    config = BigQueryExecute.Config.builder()
+            .setLocation("US").setProject("testProject").setRowAsArguments("false")
+            .setInitialRetryDuration(1L).setMaxRetryDuration(5L)
+            .setMaxRetryCount(1).setRetryMultiplier(2.0).build();
+    jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+    jobInfo = JobInfo.newBuilder(queryJobConfiguration).setJobId(jobId).build();
+    bq = new BigQueryExecute(config);
+
+    // Mock Job Creation
+    Mockito.when(bigQuery.create((JobInfo) Mockito.any())).thenReturn(queryJob);
+    Mockito.when(queryJob.waitFor()).thenReturn(queryJob);
+    Mockito.when(queryJob.getStatus()).thenReturn(jobStatus);
+    Mockito.when(jobStatus.getError()).thenReturn(bigQueryError);
+
+    // Mock Successful Query
+    Mockito.when(queryJob.getQueryResults()).thenReturn(queryResults);
+    Mockito.when(queryResults.getTotalRows()).thenReturn(1L);
+    Mockito.when(queryJob.getStatistics()).thenReturn(queryStatistics);
+    Mockito.when(queryStatistics.getTotalBytesProcessed()).thenReturn(1L);
+
+    // Mock context
+    Mockito.when(context.getMetrics()).thenReturn(stageMetrics);
+    Mockito.doNothing().when(stageMetrics).gauge(Mockito.anyString(), Mockito.anyLong());
+    Mockito.when(stageMetrics.child(Mockito.any())).thenReturn(metrics);
+    Mockito.doNothing().when(metrics).countLong(Mockito.anyString(), Mockito.anyLong());
+
+  }
+
+  @Test(expected = BigQueryJobExecutionException.class)
```

Comment:
I am still a bit confused what final sequence of exceptions would it be and, what's important, what would be a message to user. Could you please use assertThrows and validate the message as described here: https://www.baeldung.com/junit-assert-exception ? I am wondering if we would need to unwrap those exceptions to prevent unneeded layers / cluttered message

Reply:
The user will always see a `BigQueryJobExecutionException`

There are 3 cases for this to occur
1. Retry is not possible as the error reason is not in reason list, the user get the message `Failed to execute BigQuery job. Reason: {Reason from that exception}`

2. Retry was tried but failed , the user will get the message `Failed to execute BigQuery job.  io.cdap.plugin.gcp.bigquery.exception.BigQueryJobExecutionException` 
- We could provide a different message here if required, as the exception class will be  `FailsafeException` for which an additional catch could be added .

3. Someting else fails and another exception is thrown , `Failed to execute BigQuery job. Reason: {Reason from that exception}`

Reply:
I am curious about the text of the message. I want to avoid something like  `Failed to execute BigQuery job. Reason: RuntimeException: BigQueryJobExecutionException: reason`

Reply:
I have updated to assert throw, and made it more clear what will the end user see. in all cases.